src/meta.jl

```diff
@@ -0,0 +1,94 @@
+## Functions to serve metadata about the PkgServer itself
+
+function get_pkgserver_version()
+    # Get PkgServer.jl's version and git sha
+    version = Pkg.TOML.parsefile(joinpath(@__DIR__, "..", "Project.toml"))["version"]
+    try
+        repo = LibGit2.GitRepo(dirname(@__DIR__))
+        gitsha = string(LibGit2.GitHash(LibGit2.GitCommit(repo, "HEAD")))
+        return "$(version)-$(gitsha)"
+    catch
+        return "$(version)"
+    end
+end
+# We store this once, since it's not exactly going to change
+const pkgserver_version = get_pkgserver_version()
+
+function get_num_hashnamed_files(dir)
+    # If this directory doesn't exist, then we haven't cached anything!
+    if !isdir(dir)
+        return 0
+    end
+
+    num_files = 0
+    for f in readdir(dir)
+        # only consider things with a name length that is exactly a tree-hash (filters out in-progress downloads, etc...)
+        if length(f) != 40
+            continue
+        end
+
+        # Only consider actual files
+        f_path = joinpath(dir, f)
+        if !isfile(f_path)
+            continue
+        end
+
+        num_files += 1
+    end
+    return num_files
+end
+
+function get_num_pkgs_cached(cache_dir)
+    # If we don't even have a `package` directory, then don't worry about it
+    if !isdir(joinpath(cache_dir, "package"))
+        return 0
+    end
+
+    num_pkgs_cached = 0
+    for d in readdir(joinpath(cache_dir, "package"))
+        # Only consider things with a name that is exactly a UUID (filters out in-progress downloads, etc...)
+        if length(d) != 36
+            continue
+        end
+
+        # Only consider directories
+        pkg_dir = joinpath(cache_dir, "package", d)
+        if !isdir(pkg_dir)
+            continue
+        end
+
+        num_pkgs_cached += get_num_hashnamed_files(pkg_dir)
+    end
+    return num_pkgs_cached
+end
+
+function get_num_artifacts_cached(cache_dir)
+    num_artifacts_cached = get_num_hashnamed_files(joinpath(cache_dir, "artifact"))
+end
+
+function serve_meta(http::HTTP.Stream)
+    # Get a count of how many things we've cached
+    num_pkgs = get_num_pkgs_cached("cache")
+    num_artifacts = get_num_artifacts_cached("cache")
+
+    # We serve a JSON representation of some metadata about this PkgServer
+    metadata = Dict(
+        "pkgserver_version" => pkgserver_version,
+        "julia_version" => string(VERSION),
+        "packages_cached" => num_pkgs,
+        "artifacts_cached" => num_artifacts,
+    )
+
+    # We generate an extremely simple subset of JSON here
+    metadata_json = string(
+        "{",
+        join(["$(repr(k)): $(repr(v))" for (k, v) in metadata], ","),
```

Comment:
Perhaps since everything else we serve is TOML, this should also be? I know JSON is the lingua franca of HTTP services, but we don't have a JSON parser in stdlib and I don't really want to add one, whereas we do already have a TOML parser.

Reply:
I'd rather be more standard and just have whatever tool uses this have a dependency on `JSON` than depend on the fact that we have a TOML parser bundled inside of `Pkg`.  Not that I have anything against TOML, but if the only reason is because it's bundled with Julia, I don't find that all that compelling.  If you're writing a tool that does anything on the web, chances are you're going to be loading JSON at some point anyway.